features/storage/kvstore/KVStore.h

```diff
@@ -0,0 +1,199 @@
+     * @param[in]  create_flags         Flag mask.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+
+    /**
+     * @brief Get one KVStore item, given key.
+     *
+     * @param[in]  key                  Key.
+     * @param[in]  buffer               Value data buffer.
+     * @param[in]  buffer_size          Value data buffer size.
+     * @param[out] actual_size          Actual read size.
+     * @param[in]  offset               Offset to read from in data.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size = NULL, size_t offset = 0);
+
+    /**
+     * @brief Get information of a given key.
+     *
+     * @param[in]  key                  Key.
+     * @param[out] info                 Returned information structure.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int get_info(const char *key, info_t *info);
+
+    /**
+     * @brief Remove a KVStore item, given key.
+     *
+     * @param[in]  key                  Key.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int remove(const char *key);
+
+
+    /**
+     * @brief Start an incremental KVStore set sequence.
+     *
+     * @param[out] handle               Returned incremental set handle.
+     * @param[in]  key                  Key.
+     * @param[in]  final_data_size      Final value data size.
+     * @param[in]  create_flags         Flag mask.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+
+    /**
+     * @brief Add data to incremental KVStore set sequence.
+     *
+     * @param[in]  handle               Incremental set handle.
+     * @param[in]  value_data           value data to add.
+     * @param[in]  data_size            value data size.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+
+    /**
+     * @brief Finalize an incremental KVStore set sequence.
+     *
+     * @param[in]  handle               Incremental set handle.
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set_finalize(set_handle_t handle);
+
+    /**
+     * @brief Start an iteration over KVStore keys.
+     *
+     * @param[out] it                   Returned iterator handle.
+     * @param[in]  prefix               Key prefix (null for all keys).
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
```

Comment:
Are we planning on adding a KVIterator/KVSet class to allow access to these functions through C++ classes?

We could provide such a class this way (without changing any of the implementations):
``` cpp
class KVStore {
    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
    virtual int iterator_next(iterator_t it, char *key, size_t key_size);
    virtual int iterator_close(iterator_t it);

    class Iterator {
    private:
        KVStore *_kvstore;
        const char *_prefix;
        iterator_t _it;
    public:
        Iterator Iterator(KVStore *kvstore, const char *prefix = NULL)
            : _kvstore(kvstore), _prefix(prefix), _it(NULL) {
        }

        virtual int open() {
            return _kvstore->iterator_open(&_it, _prefix);
        }

        virtual int next(iterator_t it, char *key, size_t key_size) {
            return _kvstore->iterator_next(&_it, key, key_size);
        }

        virtual int close(iterator_t it) {
            return _kvstore->iterator_close(&_it);
        }
    };
}
```

Usage:
``` cpp
TDBStore kvstore;
KVStore::Iterator it(&kvstore);
char key[KVStore::MAX_KEY_SIZE+1];

kvstore.init();
it.open();
it.next(key, KVStore::MAX_KEY_SIZE);
it.close();
```

Thoughts?